armi/reactor/converters/axialExpansionChanger/axialExpansionChanger.py

```diff
@@ -420,66 +455,71 @@ def addMassToComponent(self, fromComp: "Component", toComp: "Component", delta:
-            runLog.warning(
-                f"Cannot redistribute mass from {fromComp} to {toComp} as they do not have the same nuclides.\n"
-                f"Instead, {toComp} will have it's mass changed based on the difference between its ztop and the top"
-                "of its block."
+        # limitation: fromComp and toComp **must** be the same materials.
+        if type(fromComp.material) is not type(toComp.material):
+            raise RuntimeError(
+                f"Cannot redistribute mass from {fromComp} to {toComp} as they are not the same material.\n"
+                f"{type(fromComp.material)} is not {type(toComp.material)}"
             )
-            if delta > 0:
-                # expansion, add mass to toComp
-                return self._noRedistribution(toComp, delta)
-            else:
-                # contraction, add mass to fromComp
-                return self._noRedistribution(fromComp, delta)
 
-        delta = abs(delta)
-        ## calculate new number densities for each isotope based on the expected total mass
         toCompVolume = toComp.getArea() * toComp.height
-        fromCompVolume = fromComp.getArea() * delta
+        fromCompVolume = fromComp.getArea() * abs(delta)
         newVolume = fromCompVolume + toCompVolume
 
-        ## calculate the mass of each nuclide
-        massByNucFrom = {}
-        massByNucTo = {}
-        for nuc in nucsFrom:
-            massByNucFrom[nuc] = densityTools.getMassInGrams(nuc, fromCompVolume, fromComp.getNumberDensity(nuc))
-            massByNucTo[nuc] = densityTools.getMassInGrams(nuc, toCompVolume, toComp.getNumberDensity(nuc))
-
-        ## calculate the ndens from the new mass
+        ## calculate the mass of each nuclide and then the ndens for the new mass
         newNDens: dict[str, float] = {}
-        for nuc in nucsFrom:
-            newNDens[nuc] = densityTools.calculateNumberDensity(nuc, massByNucFrom[nuc] + massByNucTo[nuc], newVolume)
+        for nuc in fromComp.getNuclides():
+            massByNucFrom = densityTools.getMassInGrams(nuc, fromCompVolume, fromComp.getNumberDensity(nuc))
+            massByNucTo = densityTools.getMassInGrams(nuc, toCompVolume, toComp.getNumberDensity(nuc))
+            newNDens[nuc] = densityTools.calculateNumberDensity(nuc, massByNucFrom + massByNucTo, newVolume)
 
         ## Set newNDens on toComp
         toComp.setNumberDensities(newNDens)
 
-    def _noRedistribution(self, c: "Component", delta: float):
-        """Calculate new number densities for each isotope based on the expected total mass.
+        # calculate the new temperature of toComp.
+        if fromComp.temperatureInC == toComp.temperatureInC:
+            newToCompTemp = toComp.temperatureInC
+        else:
+            rhs = newVolume / (toComp.height + abs(delta))
+            newToCompTemp = brentq(
+                f=lambda T: toComp.getArea(Tc=T) - rhs, a=fromComp.temperatureInC, b=toComp.temperatureInC
```

Comment:
might break ruff but
```suggestion
            targetArea = newVolume / (toComp.height + abs(delta))
            newToCompTemp = brentq(
                f=lambda T: toComp.getArea(Tc=T) - targetArea , 
                a=fromComp.temperatureInC, 
                b=toComp.temperatureInC
```

Reply:
https://github.com/terrapower/armi/pull/2198/commits/38aae19060d6446b57834eb694ad809656524a71